Installation issue: elpa+cuda

### Steps to reproduce the issue

```console
$ spack spec -I "elpa+cuda+mpi cuda_arch=80 ^[virtuals=scalapack] netlib-scalapack ^[virtuals=blas,lapack] openblas threads=openmp"
Input spec
--------------------------------
 -   elpa+cuda+mpi cuda_arch=80
 -       ^netlib-scalapack
 -       ^openblas threads=openmp

Concretized
--------------------------------
 -   elpa@2023.11.001-patched%gcc@11.4.0~autotune+cuda~gpu_streams+mpi+openmp~rocm build_system=autotools cuda_arch=80 arch=linux-ubuntu23.10-zen4
[+]      ^cuda@11.8.0%gcc@11.4.0~allow-unsupported-compilers~dev build_system=generic arch=linux-ubuntu23.10-zen4
[+]          ^libxml2@2.10.3%gcc@11.4.0+pic~python+shared build_system=autotools arch=linux-ubuntu23.10-zen4
[+]              ^libiconv@1.17%gcc@11.4.0 build_system=autotools libs=shared,static arch=linux-ubuntu23.10-zen4
[+]      ^gcc-runtime@11.4.0%gcc@11.4.0 build_system=generic arch=linux-ubuntu23.10-zen4
[+]      ^gmake@4.4.1%gcc@11.4.0~guile build_system=generic arch=linux-ubuntu23.10-zen4
[+]      ^libtool@2.4.7%gcc@11.4.0 build_system=autotools arch=linux-ubuntu23.10-zen4
[+]          ^findutils@4.9.0%gcc@11.4.0 build_system=autotools patches=440b954 arch=linux-ubuntu23.10-zen4
[+]          ^m4@1.4.19%gcc@11.4.0+sigsegv build_system=autotools patches=9dc5fbd,bfdffa7 arch=linux-ubuntu23.10-zen4
[+]              ^diffutils@3.10%gcc@11.4.0 build_system=autotools arch=linux-ubuntu23.10-zen4
[+]              ^libsigsegv@2.14%gcc@11.4.0 build_system=autotools arch=linux-ubuntu23.10-zen4
[+]      ^mpich@4.1.2%gcc@11.4.0~argobots~cuda+fortran+hwloc+hydra+libxml2+pci~rocm+romio~slurm~verbs+wrapperrpath~xpmem build_system=autotools datatype-engine=auto device=ch3 netmod=tcp pmi=pmi arch=linux-ubuntu23.10-zen4
[+]          ^hwloc@2.9.1%gcc@11.4.0~cairo~cuda~gl~libudev+libxml2~netloc~nvml~oneapi-level-zero~opencl+pci~rocm build_system=autotools libs=shared,static arch=linux-ubuntu23.10-zen4
[+]          ^libpciaccess@0.17%gcc@11.4.0 build_system=autotools arch=linux-ubuntu23.10-zen4
[+]              ^util-macros@1.19.3%gcc@11.4.0 build_system=autotools arch=linux-ubuntu23.10-zen4
[+]          ^pkgconf@1.9.5%gcc@11.4.0 build_system=autotools arch=linux-ubuntu23.10-zen4
[+]      ^netlib-scalapack@2.2.0%gcc@11.4.0~ipo~pic+shared build_system=cmake build_type=Release generator=make patches=072b006,1c9ce5f,244a9aa arch=linux-ubuntu23.10-zen4
[+]          ^cmake@3.27.9%gcc@11.4.0~doc+ncurses+ownlibs build_system=generic build_type=Release arch=linux-ubuntu23.10-zen4
[+]              ^curl@8.6.0%gcc@11.4.0~gssapi~ldap~libidn2~librtmp~libssh~libssh2+nghttp2 build_system=autotools libs=shared,static tls=openssl arch=linux-ubuntu23.10-zen4
[+]                  ^nghttp2@1.57.0%gcc@11.4.0 build_system=autotools arch=linux-ubuntu23.10-zen4
[+]      ^openblas@0.3.26%gcc@11.4.0~bignuma~consistent_fpcsr+dynamic_dispatch+fortran~ilp64+locking+pic+shared build_system=makefile symbol_suffix=none threads=openmp arch=linux-ubuntu23.10-zen4
[+]          ^perl@5.38.0%gcc@11.4.0+cpanm+opcode+open+shared+threads build_system=generic patches=714e4d1 arch=linux-ubuntu23.10-zen4
[+]              ^berkeley-db@18.1.40%gcc@11.4.0+cxx~docs+stl build_system=autotools patches=26090f4,b231fcc arch=linux-ubuntu23.10-zen4
[+]      ^python@3.11.7%gcc@11.4.0+bz2+crypt+ctypes+dbm~debug+libxml2+lzma~nis~optimizations+pic+pyexpat+pythoncmd+readline+shared+sqlite3+ssl~tkinter+uuid+zlib build_system=generic patches=13fa8bf,b0615b2,ebdca64,f2fd060 arch=linux-ubuntu23.10-zen4
[+]          ^bzip2@1.0.8%gcc@11.4.0~debug~pic+shared build_system=generic arch=linux-ubuntu23.10-zen4
[+]          ^expat@2.6.2%gcc@11.4.0+libbsd build_system=autotools arch=linux-ubuntu23.10-zen4
[+]              ^libbsd@0.12.1%gcc@11.4.0 build_system=autotools arch=linux-ubuntu23.10-zen4
[+]                  ^libmd@1.0.4%gcc@11.4.0 build_system=autotools arch=linux-ubuntu23.10-zen4
[+]          ^gdbm@1.23%gcc@11.4.0 build_system=autotools arch=linux-ubuntu23.10-zen4
[+]          ^gettext@0.22.4%gcc@11.4.0+bzip2+curses+git~libunistring+libxml2+pic+shared+tar+xz build_system=autotools arch=linux-ubuntu23.10-zen4
[+]              ^tar@1.34%gcc@11.4.0 build_system=autotools zip=pigz arch=linux-ubuntu23.10-zen4
[+]                  ^pigz@2.8%gcc@11.4.0 build_system=makefile arch=linux-ubuntu23.10-zen4
[+]                  ^zstd@1.5.5%gcc@11.4.0+programs build_system=makefile compression=none libs=shared,static arch=linux-ubuntu23.10-zen4
[+]          ^libffi@3.4.4%gcc@11.4.0 build_system=autotools patches=070b1f3 arch=linux-ubuntu23.10-zen4
[+]          ^libxcrypt@4.4.35%gcc@11.4.0~obsolete_api build_system=autotools patches=4885da3 arch=linux-ubuntu23.10-zen4
[+]          ^ncurses@6.4%gcc@11.4.0~symlinks+termlib abi=none build_system=autotools patches=7a351bc arch=linux-ubuntu23.10-zen4
[+]          ^openssl@3.2.1%gcc@11.4.0~docs+shared build_system=generic certs=mozilla arch=linux-ubuntu23.10-zen4
[+]              ^ca-certificates-mozilla@2023-05-30%gcc@11.4.0 build_system=generic arch=linux-ubuntu23.10-zen4
[+]          ^readline@8.2%gcc@11.4.0 build_system=autotools patches=bbf97f1 arch=linux-ubuntu23.10-zen4
[+]          ^sqlite@3.43.2%gcc@11.4.0+column_metadata+dynamic_extensions+fts~functions+rtree build_system=autotools arch=linux-ubuntu23.10-zen4
[+]          ^util-linux-uuid@2.38.1%gcc@11.4.0 build_system=autotools arch=linux-ubuntu23.10-zen4
[+]          ^xz@5.4.6%gcc@11.4.0~pic build_system=autotools libs=shared,static arch=linux-ubuntu23.10-zen4
[+]          ^zlib-ng@2.1.6%gcc@11.4.0+compat+new_strategies+opt+pic+shared build_system=autotools arch=linux-ubuntu23.10-zen4
```


### Error message

<details><summary>Error message</summary>

<pre>
==> Installing elpa-2023.11.001-patched-zemwzxgkynnfvzyvjppc7zzgmtxgixf7 [32/32]
==> No binary for elpa-2023.11.001-patched-zemwzxgkynnfvzyvjppc7zzgmtxgixf7 found: installing from source
==> Using cached archive: /home/rmeli/spack/var/spack/cache/_source-cache/archive/62/62ee109afc06539507f459c08b958dc4db65b757dbd77f927678c77f7687415e.tar.gz
==> No patches needed for elpa
==> elpa: Executing phase: 'autoreconf'
==> elpa: Executing phase: 'configure'
==> elpa: Executing phase: 'build'
==> Error: ProcessError: Command exited with status 2:
    'make' '-j1' 'V=1'

2 errors found in build log:
     27016    mv -f $depbase.Tpo $depbase.Plo
     27017    libtool: compile:  /home/rmeli/spack/opt/spack/linux-ubuntu23.10-zen4/gcc-11.4.0/mpich-4.1.2-2yjnzcd4y34l5spntatff5e75i5n6dlr/bin/mpicc -DHAVE_CONFIG_H 
              -I. -I/tmp/rmeli/spack-stage/spack-stage-elpa-2023.11.001-patched-zemwzxgkynnfvzyvjppc7zzgmtxgixf7/spack-src -O3 -MT src/ftimings/papi.lo -MD -MP -MF sr
              c/ftimings/.deps/papi.Tpo -c /tmp/rmeli/spack-stage/spack-stage-elpa-2023.11.001-patched-zemwzxgkynnfvzyvjppc7zzgmtxgixf7/spack-src/src/ftimings/papi.c 
               -fPIC -DPIC -o src/ftimings/.libs/papi.o
     27018    libtool: compile:  /home/rmeli/spack/opt/spack/linux-ubuntu23.10-zen4/gcc-11.4.0/mpich-4.1.2-2yjnzcd4y34l5spntatff5e75i5n6dlr/bin/mpicc -DHAVE_CONFIG_H 
              -I. -I/tmp/rmeli/spack-stage/spack-stage-elpa-2023.11.001-patched-zemwzxgkynnfvzyvjppc7zzgmtxgixf7/spack-src -O3 -MT src/ftimings/papi.lo -MD -MP -MF sr
              c/ftimings/.deps/papi.Tpo -c /tmp/rmeli/spack-stage/spack-stage-elpa-2023.11.001-patched-zemwzxgkynnfvzyvjppc7zzgmtxgixf7/spack-src/src/ftimings/papi.c 
              -o src/ftimings/papi.o >/dev/null 2>&1
     27019    NVCC="nvcc" /bin/bash ./libtool --mode=compile --tag=CC /tmp/rmeli/spack-stage/spack-stage-elpa-2023.11.001-patched-zemwzxgkynnfvzyvjppc7zzgmtxgixf7/spa
              ck-src/nvcc_wrap -arch sm_80 -O2 -I/home/rmeli/spack/opt/spack/linux-ubuntu23.10-zen4/gcc-11.4.0/cuda-11.8.0-i2phxppqbczp4p7jg54aquu3id3yipjl/include   
              -L/home/rmeli/spack/opt/spack/linux-ubuntu23.10-zen4/gcc-11.4.0/openblas-0.3.26-6xjkrifjo4tws5427usxw4l2573yqhtx/lib -L/home/rmeli/spack/opt/spack/linux
              -ubuntu23.10-zen4/gcc-11.4.0/openblas-0.3.26-6xjkrifjo4tws5427usxw4l2573yqhtx/lib -fopenmp -L/home/rmeli/spack/opt/spack/linux-ubuntu23.10-zen4/gcc-11.4
              .0/cuda-11.8.0-i2phxppqbczp4p7jg54aquu3id3yipjl/lib64 -I./ -I/tmp/rmeli/spack-stage/spack-stage-elpa-2023.11.001-patched-zemwzxgkynnfvzyvjppc7zzgmtxgixf
              7/spack-src/  -c /tmp/rmeli/spack-stage/spack-stage-elpa-2023.11.001-patched-zemwzxgkynnfvzyvjppc7zzgmtxgixf7/spack-src/src/GPU/CUDA/elpa_index_nvidia_g
              pu.cu -o src/GPU/CUDA/elpa_index_nvidia_gpu.lo
     27020    libtool: compile:  /tmp/rmeli/spack-stage/spack-stage-elpa-2023.11.001-patched-zemwzxgkynnfvzyvjppc7zzgmtxgixf7/spack-src/nvcc_wrap -arch sm_80 -O2 -I/h
              ome/rmeli/spack/opt/spack/linux-ubuntu23.10-zen4/gcc-11.4.0/cuda-11.8.0-i2phxppqbczp4p7jg54aquu3id3yipjl/include -L/home/rmeli/spack/opt/spack/linux-ubu
              ntu23.10-zen4/gcc-11.4.0/openblas-0.3.26-6xjkrifjo4tws5427usxw4l2573yqhtx/lib -L/home/rmeli/spack/opt/spack/linux-ubuntu23.10-zen4/gcc-11.4.0/openblas-0
              .3.26-6xjkrifjo4tws5427usxw4l2573yqhtx/lib -fopenmp -L/home/rmeli/spack/opt/spack/linux-ubuntu23.10-zen4/gcc-11.4.0/cuda-11.8.0-i2phxppqbczp4p7jg54aquu3
              id3yipjl/lib64 -I./ -I/tmp/rmeli/spack-stage/spack-stage-elpa-2023.11.001-patched-zemwzxgkynnfvzyvjppc7zzgmtxgixf7/spack-src/ -c /tmp/rmeli/spack-stage/
              spack-stage-elpa-2023.11.001-patched-zemwzxgkynnfvzyvjppc7zzgmtxgixf7/spack-src/src/GPU/CUDA/elpa_index_nvidia_gpu.cu  -fPIC -DPIC -o src/GPU/CUDA/.libs
              /elpa_index_nvidia_gpu.o
     27021    nvcc fatal   : Unknown option '-fopenmp'
  >> 27022    make[1]: *** [Makefile:92133: src/GPU/CUDA/elpa_index_nvidia_gpu.lo] Error 1
     27023    make[1]: Leaving directory '/tmp/rmeli/spack-stage/spack-stage-elpa-2023.11.001-patched-zemwzxgkynnfvzyvjppc7zzgmtxgixf7/spack-src/spack-build'
  >> 27024    make: *** [Makefile:68495: all] Error 2

See build log for details:
  /tmp/rmeli/spack-stage/spack-stage-elpa-2023.11.001-patched-zemwzxgkynnfvzyvjppc7zzgmtxgixf7/spack-build-out.txt

</pre></details>


### Information on your system

* **Spack:** 0.22.0.dev0 (b362362291816679873d5f2dad37a31748d7e42c)
* **Python:** 3.10.12
* **Platform:** linux-ubuntu23.10-zen4
* **Concretizer:** clingo

or

* **Spack:** 0.22.0.dev0 (cf863f1c2ee0cac496ada9cd1bd4f660836b67a6)
* **Python:** 3.6.15
* **Platform:** linux-sles15-neoverse_v2
* **Concretizer:** clingo

### Additional information

[spack-build-env.txt](https://github.com/spack/spack/files/15151615/spack-build-env.txt)
[spack-build-out.txt](https://github.com/spack/spack/files/15151617/spack-build-out.txt)

Tagging @amd-toolchain-support because this particular issue appears to be related to https://github.com/spack/spack/pull/42747.


### General information

- [X] I have run `spack debug report` and reported the version of Spack/Python/Platform
- [X] I have run `spack maintainers <name-of-the-package>` and **@mentioned** any maintainers
- [X] I have uploaded the build log and environment files
- [X] I have searched the issues of this repo and believe this is not a duplicate

--- a/var/spack/repos/builtin/packages/elpa/package.py
+++ b/var/spack/repos/builtin/packages/elpa/package.py
@@ -54,7 +54,7 @@ class Elpa(AutotoolsPackage, CudaPackage, ROCmPackage):
     variant(
         "autotune",
         default=False,
-        when="@2021.11.01",
+        when="@2021.11.01:",
         description="Enables autotuning for matrix restribution",
     )
     variant(

--- a/var/spack/repos/builtin/packages/elpa/package.py
+++ b/var/spack/repos/builtin/packages/elpa/package.py
@@ -25,6 +25,9 @@ class Elpa(AutotoolsPackage, CudaPackage, ROCmPackage):
     version("master", branch="master")
 
     version(
+        "2024.03.001", sha256="41c6cbf56d2dac26443faaba8a77307d261bf511682a64b96e24def77c813622"
+    )
+    version(
         "2023.11.001-patched",
         sha256="62ee109afc06539507f459c08b958dc4db65b757dbd77f927678c77f7687415e",
         url="https://elpa.mpcdf.mpg.de/software/tarball-archive/Releases/2023.11.001/elpa-2023.11.001-patched.tar.gz",
@@ -48,11 +51,22 @@ class Elpa(AutotoolsPackage, CudaPackage, ROCmPackage):
 
     variant("openmp", default=True, description="Activates OpenMP support")
     variant("mpi", default=True, description="Activates MPI support")
-    variant("gpu_streams", default=True, description="Activates GPU streams support")
+    variant(
+        "autotune",
+        default=False,
+        when="@2021.11.01",
+        description="Enables autotuning for matrix restribution",
+    )
+    variant(
+        "gpu_streams",
+        default=True,
+        when="@2021.11.001: +cuda",
+        description="Activates GPU streams support",
+    )
 
     patch("fujitsu.patch", when="%fj")
 
-    depends_on("autoconf", type="build", when="@master")
+    depends_on("autoconf@2.71:", type="build", when="@master")
     depends_on("automake", type="build", when="@master")
 
     depends_on("blas")
@@ -62,12 +76,20 @@ class Elpa(AutotoolsPackage, CudaPackage, ROCmPackage):
     depends_on("rocblas", when="+rocm")
     depends_on("libtool", type="build")
     depends_on("python@3:", type="build")
+    depends_on("scalapack", when="+autotune")
 
-    with when("@2021.11.01:"):
-        variant(
-            "autotune", default=False, description="Enables autotuning for matrix restribution"
+    # Force openmp propagation on some providers of blas/lapack, as adviced by docs
+    # https://gitlab.mpcdf.mpg.de/elpa/elpa/-/blob/master/documentation/PERFORMANCE_TUNING.md?ref_type=heads#builds-with-openmp-enabled
+    with when("+openmp"):
+        requires("^openblas threads=openmp", when="^[virtuals=blas,lapack] openblas")
+        requires("^intel-mkl threads=openmp", when="^[virtuals=blas,lapack] intel-mkl")
+        requires(
+            "^intel-oneapi-mkl threads=openmp", when="^[virtuals=blas,lapack] intel-oneapi-mkl"
         )
-        depends_on("scalapack", when="+autotune")
+        requires(
+            "^intel-parallel-studio threads=openmp",
+            when="^[virtuals=blas,lapack] intel-parallel-studio",
+        )
 
     # fails to build due to broken type-bound procedures in OMP parallel regions
     conflicts(
@@ -76,10 +98,9 @@ class Elpa(AutotoolsPackage, CudaPackage, ROCmPackage):
         msg="ELPA-2021.05.001+ requires GCC-8+ for OpenMP support",
     )
     conflicts("+mpi", when="+rocm", msg="ROCm support and MPI are not yet compatible")
-
     conflicts(
         "+gpu_streams",
-        when="+openmp",
+        when="@:2023.11.001-patched +openmp",
         msg="GPU streams currently not supported in combination with OpenMP",
     )
 
@@ -193,13 +214,8 @@ class Elpa(AutotoolsPackage, CudaPackage, ROCmPackage):
         options += self.enable_or_disable("openmp")
 
         # Additional linker search paths and link libs
-        ldflags = [spec["blas"].libs.search_flags, spec["lapack"].libs.search_flags]
+        ldflags = [spec["blas"].libs.search_flags, spec["lapack"].libs.search_flags, "-lstdc++"]
         libs = [spec["lapack"].libs.link_flags, spec["blas"].libs.link_flags]
-
-        # If using blas with openmp support, link with openmp
-        # Needed for Spack-provided OneAPI MKL and for many externals
-        if self.spec["blas"].satisfies("threads=openmp"):
-            ldflags.append(self.compiler.openmp_flag)
 
         options += [f'LDFLAGS={" ".join(ldflags)}', f'LIBS={" ".join(libs)}']
 

--- a/var/spack/repos/builtin/packages/elpa/package.py
+++ b/var/spack/repos/builtin/packages/elpa/package.py
@@ -52,7 +52,7 @@ class Elpa(AutotoolsPackage, CudaPackage, ROCmPackage):
     variant("openmp", default=True, description="Activates OpenMP support")
     variant("mpi", default=True, description="Activates MPI support")
 
-    with when("@2021.11.01:"):
+    with when("@2021.11.001:"):
         variant(
             "autotune", default=False, description="Enables autotuning for matrix restribution"
         )

--- a/var/spack/repos/builtin/packages/elpa/package.py
+++ b/var/spack/repos/builtin/packages/elpa/package.py
@@ -51,18 +51,14 @@ class Elpa(AutotoolsPackage, CudaPackage, ROCmPackage):
 
     variant("openmp", default=True, description="Activates OpenMP support")
     variant("mpi", default=True, description="Activates MPI support")
-    variant(
-        "autotune",
-        default=False,
-        when="@2021.11.01:",
-        description="Enables autotuning for matrix restribution",
-    )
-    variant(
-        "gpu_streams",
-        default=True,
-        when="@2021.11.001: +cuda",
-        description="Activates GPU streams support",
-    )
+
+    with when("@2021.11.01:"):
+        variant(
+            "autotune", default=False, description="Enables autotuning for matrix restribution"
+        )
+        variant(
+            "gpu_streams", default=True, when="+cuda", description="Activates GPU streams support"
+        )
 
     patch("fujitsu.patch", when="%fj")
 

